futures/src/supervisor.rs

```diff
@@ -0,0 +1,843 @@
+//! Lightweight supervisor for tokio tasks.
+//!
+//! What it does:
+//!
+//! - Monitors multiple children (as spawned [`JoinHandle`])
+//! - Provides a single shutdown signal for everything
+//! - Supports graceful shutdown timeout before aborting a child (via [`OnShutdown`])
+//! - If a child panics, initiates shutdown and exits with an error
+//! - If a child exits before shutdown signal, also initiates shutdown and exits with an error.
+//!   Note: this might not be always the desirable behaviour, but _currently_ there are no other
+//!   cases in Iroha.
+//!   This behaviour could be easily extended to support refined strategies.
+//! - Logs children's lifecycle
+//!
+//! What it doesn't:
+//!
+//! - Doesn't support restarting child.
+//!   To implement that, we need a formal actor system.
```

Comment:
Simply put, I mean an actor system from Erlang =) I would call it a set of well-put-together design choices.

In this case,  we need a registry of actor addresses to allow indirect communication. It is necessary for restarts, so that actor A, communicating with actor B, can get an address of actor B from the registry instead of storing the address directly. This way, actor B might fail, restart, and update its address in the registry, and it happens seamlessly for actor A and for the rest of the system. Here is more on what I mean: [Registry — Elixir](https://hexdocs.pm/elixir/1.17.0-rc.1/Registry.html)

Extended the comment.

Reply:
Iroha used to have uniform actor system but it was kinda clunky to put all actors under the same umbrella and had some nasty bugs due to this registry.

Reply:
Well, we probably don't need it that much.